# Case Study

In [1]:
#importing libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [6]:
dataset = pd.read_csv("StudentsPerformance.csv")
dataset

gender race/ethnicity parental level of education         lunch  \
0    female        group C            some high school  free/reduced   
1    female        group B                 high school  free/reduced   
2    female        group B            some high school  free/reduced   
3    female        group B                some college      standard   
4    female        group C                some college  free/reduced   
..      ...            ...                         ...           ...   
995    male        group E                some college      standard   
996    male        group A                some college      standard   
997    male        group D                some college      standard   
998    male        group E          associate's degree  free/reduced   
999    male        group E           bachelor's degree      standard   

    test preparation course  math score  reading score  writing score  
0                      none           0             17             10  
1                      none           8             24             23  
2                      none          18             32             28  
3                      none          11             38             32  
4                      none          22             39             33  
..                      ...         ...            ...            ...  
995               completed          99             87             81  
996               completed         100             96             86  
997               completed         100             97             99  
998               completed         100            100             93  
999               completed         100            100            100  

[1000 rows x 8 columns]

In [7]:
dataset.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group C            some high school  free/reduced   
1  female        group B                 high school  free/reduced   
2  female        group B            some high school  free/reduced   
3  female        group B                some college      standard   
4  female        group C                some college  free/reduced   

  test preparation course  math score  reading score  writing score  
0                    none           0             17             10  
1                    none           8             24             23  
2                    none          18             32             28  
3                    none          11             38             32  
4                    none          22             39             33

In [8]:
dataset.tail()

gender race/ethnicity parental level of education         lunch  \
995   male        group E                some college      standard   
996   male        group A                some college      standard   
997   male        group D                some college      standard   
998   male        group E          associate's degree  free/reduced   
999   male        group E           bachelor's degree      standard   

    test preparation course  math score  reading score  writing score  
995               completed          99             87             81  
996               completed         100             96             86  
997               completed         100             97             99  
998               completed         100            100             93  
999               completed         100            100            100

In [9]:
dataset.isnull().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

In [28]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


Given dataset contains details on 1000 students identified only by their gender and race and tabulised along with details of mark score,test preparation course etc. 
Dataset is evaluated for null values and is devoid of any.

### 1. To find out how many males and females participated in the test. 

In [17]:
#to find total of each gender participated.
number_gender = dataset['gender'].value_counts()
number_gender

female    518
male      482
Name: gender, dtype: int64

518 females and 482 males participated in the test.

### 2. Students' parental level of education 

In [18]:
dataset['parental level of education'].unique()

array(['some high school', 'high school', 'some college',
       "associate's degree", "bachelor's degree", "master's degree"],
      dtype=object)

In [21]:
education_parents = dataset['parental level of education'].value_counts()
education_parents

some college          226
associate's degree    222
high school           196
some high school      179
bachelor's degree     118
master's degree        59
Name: parental level of education, dtype: int64

There are 6 different classes to which parents of students belongs in which parents belonging to college level of education is greater in number. Parents having master's degree is lesser among the list with only 59 in number in a pool of 1000 members.

### 3. Who scores the most based on average.. 

#### w.r.t gender 

In [25]:
#for math
dataset[['gender','math score']].groupby('gender').mean().sort_values('math score',ascending=False)

math score
gender            
male     68.821577
female   63.376448

In [26]:
#for reading score
dataset[['gender','reading score']].groupby('gender').mean().sort_values('reading score',ascending=False)

reading score
gender               
female      72.590734
male        65.545643

In [27]:
#for writing score
dataset[['gender','writing score']].groupby('gender').mean().sort_values('writing score',ascending=False)

writing score
gender               
female      72.467181
male        63.446058

On calculating averages among students wrt gender female has greater mean in both reading and writing scores.
But males show greater mean value than females in math score.

*Female in reading and writin scores and male in math score have higher marks*

#### w.r.t test preparation course 

In [53]:
#for math
dataset[['test preparation course','math score']].groupby('test preparation course').mean().sort_values('math score',ascending=False)

math score
test preparation course            
completed                  69.96648
none                       63.78972

In [54]:
#for reading score
dataset[['test preparation course','reading score']].groupby('test preparation course').mean().sort_values('reading score',ascending=False)

reading score
test preparation course               
completed                    74.175978
none                         66.417445

In [55]:
#for writing score
dataset[['test preparation course','writing score']].groupby('test preparation course').mean().sort_values('writing score',ascending=False)

writing score
test preparation course               
completed                    74.684358
none                         64.457944

On calculating averages among students wrt test preparation course,we can get a clear idea that for all category of scores, those who have completed the course is having higher marks when compared to the students who haven't completed their course.

*So there is a clear cut advance for those who have completed the course,since they scored higher with great difference from others*

### 4. Scoring variations 

#### w.r.t gender 

In [59]:
#for math
dataset[['gender','math score']].groupby('gender').std().sort_values('math score')

math score
gender            
male     14.556411
female   16.029928

In [60]:
#for reading score
dataset[['gender','reading score']].groupby('gender').std().sort_values('reading score')

reading score
gender               
male        14.149594
female      14.411018

In [61]:
#for writing score
dataset[['gender','writing score']].groupby('gender').std().sort_values('writing score')

writing score
gender               
male        14.227225
female      14.844842

Computed standard deviation wrt gender shows almost same variation for both male and female in the case of reading and writing scores while there is a slight variation in math scores between male and female.

#### w.r.t test preparation course 

In [62]:
#for math
dataset[['test preparation course','math score']].groupby('test preparation course').std().sort_values('math score')

math score
test preparation course            
completed                 14.521847
none                      15.705689

In [64]:
#for reading score
dataset[['test preparation course','reading score']].groupby('test preparation course').std().sort_values('reading score')

reading score
test preparation course               
completed                    13.537572
none                         14.608896

In [65]:
#for writing score
dataset[['test preparation course','writing score']].groupby('test preparation course').std().sort_values('writing score')

writing score
test preparation course               
completed                    13.236412
none                         15.041667

Variation study done by finding standard deviation wrt test preparation course gives small variations.
But the point to be noted is that *variation in marks is certainly less for those who have completed the test preparation course* when compared to those who haven't completed their course.

### 5. Top 25% students based on  math score

In order to find top 25% students based on math score we can consider the quartile method.
Quartiles divide data into 4 equal parts and the best one-forth of total students will be the ones who are above the 3rd quartile mark, that is above 75 percentile.

In [68]:
#to get the third quartile mark
dataset.describe()

math score  reading score  writing score
count  1000.000000    1000.000000    1000.000000
mean     66.001000      69.195000      68.119000
std      15.569567      14.706436      15.226381
min       0.000000      17.000000      10.000000
25%      57.000000      59.000000      58.000000
50%      66.000000      70.000000      69.000000
75%      77.000000      79.250000      79.000000
max     100.000000     100.000000     100.000000

Above table shows that, students having math score of 77 and above are the top 25% in the class wrt math score.

In [79]:
#list of top 25% students as needed to award bonus points
toppers = dataset[dataset["math score"]>=77]

pd.set_option('display.max_rows',255)

toppers

gender race/ethnicity parental level of education         lunch  \
414  female        group A                 high school  free/reduced   
415  female        group B             master's degree  free/reduced   
416  female        group B           bachelor's degree  free/reduced   
417  female        group B             master's degree      standard   
418  female        group B                 high school      standard   
419  female        group C           bachelor's degree      standard   
420  female        group C           bachelor's degree      standard   
421  female        group C                some college  free/reduced   
422  female        group C            some high school      standard   
423  female        group C            some high school      standard   
424  female        group D          associate's degree  free/reduced   
425  female        group D             master's degree      standard   
426  female        group D                some college      standard   
427  female        group D          associate's degree      standard   
428  female        group D                some college  free/reduced   
429  female        group E            some high school      standard   
430  female        group A                some college      standard   
431  female        group B           bachelor's degree  free/reduced   
432  female        group D           bachelor's degree      standard   
433  female        group D                 high school      standard   
434  female        group D           bachelor's degree  free/reduced   
435  female        group D             master's degree      standard   
436  female        group B                some college      standard   
437  female        group C           bachelor's degree      standard   
438  female        group D                some college      standard   
439  female        group D           bachelor's degree      standard   
440  female        group D                some college  free/reduced   
441  female        group D                some college      standard   
442  female        group E          associate's degree      standard   
443  female        group E           bachelor's degree      standard   
444  female        group B          associate's degree      standard   
445  female        group D                some college      standard   
446  female        group D            some high school      standard   
447  female        group D            some high school      standard   
448  female        group E           bachelor's degree      standard   
449  female        group E            some high school      standard   
450  female        group B                 high school      standard   
451  female        group C             master's degree      standard   
452  female        group C           bachelor's degree      standard   
453  female        group C          associate's degree      standard   
454  female        group C                 high school      standard   
455  female        group D            some high school      standard   
456  female        group E             master's degree      standard   
457  female        group E             master's degree  free/reduced   
458  female        group A          associate's degree      standard   
459  female        group B          associate's degree      standard   
460  female        group B            some high school      standard   
461  female        group B                some college      standard   
462  female        group C                some college      standard   
463  female        group C          associate's degree  free/reduced   
464  female        group D                some college      standard   
465  female        group D          associate's degree      standard   
466  female        group E          associate's degree      standard   
467  female        group C           bachelor's degree      standard   
468  female        group C          associate

*Above is the list of 255 students  who are eligible for bonus points based on math score*